pkg/api/authz.go

```diff
@@ -0,0 +1,222 @@
+package api
+
+import (
+	"context"
+	"encoding/base64"
+	"net/http"
+	"strings"
+	"time"
+
+	"github.com/anuvu/zot/pkg/log"
+	"github.com/gorilla/mux"
+)
+
+type key int
+
+const (
+	CREATE           = "create"
+	READ             = "read"
+	UPDATE           = "update"
+	DELETE           = "delete"
+	contextKeyID key = 0
+)
+
+type AccessControlConfig struct {
+	Repositories Repositories
+	AdminPolicy  Policy
+}
+
+type Repositories map[string]PolicyGroup
+
+type Repository struct {
+	Name          string
+	Policies      []Policy
+	DefaultPolicy []string
+}
+
+type PolicyGroup struct {
+	Policies      []Policy
+	DefaultPolicy []string
+}
+
+type Policy struct {
+	Users   []string
+	Actions []string
+}
+
+// AccessController authorizes users to act on resources.
+type AccessController struct {
+	Config *AccessControlConfig
+	Log    log.Logger
+}
+
+// AccessControllerContext context passed down to http.Handlers.
+type AccessControllerContext struct {
+	userAllowedRepos []string
+	isAdmin          bool
+}
+
+func NewAccessController(config *Config) *AccessController {
+	return &AccessController{
+		Config: config.AccessControl,
+		Log:    log.NewLogger(config.Log.Level, config.Log.Output),
+	}
+}
+
+// getReadRepos get repositories from config file that the user has READ perms.
+func (a *AccessController) getReadRepos(username string) []string {
```

Comment:
This method result will be used to search a repo only, if yes can we make its return type hashset ?

Reply:
I'm not sure I understand.
This method is used to search all repos on which the user has READ perm.
The result is used to filter out zot's repos by username's rights.

Can you detail a bit?